src/utils/timeout.ts

```diff
@@ -1,3 +1,18 @@
 export async function timeout(ms: number): Promise<void> {
   return new Promise((resolve) => setTimeout(resolve, ms));
 }
+
+export class TimeoutError extends Error {}
+
+export const promiseTimeout = async <T>(ms: number, promise: Promise<T>): Promise<T> => {
```

Comment:
please add simple 2 unit test of race between the above timeout (from line 1) vs ms
one where the promise is suppose to win and the other where the ms value should win